parser_library/src/lsp/completion_item.cpp

```diff
@@ -32,214 +32,263 @@ completion_item_s::completion_item_s(std::string label,
+        }
+        else if (no_optional == 2 && mach_operands.size() - i == 2)
+        {
+            curr_op_with_mnem += "[";
+            curr_op_without_mnem += "[";
+            if (i != 0)
+                curr_op_with_mnem += ",";
+            if (!first)
+                curr_op_without_mnem += ",";
+            curr_op_with_mnem += mach_operands[i].to_string() + "[,";
+            curr_op_without_mnem += mach_operands[i].to_string() + "[,";
+        }
+        subs_ops_mnems << curr_op_with_mnem;
+        subs_ops_nomnems << curr_op_without_mnem;
+        first = false;
+    }
+    detail_ss << "Operands: " + subs_ops_nomnems.str();
+    doc_ss << "Mnemonic code for " << mnemonic_instr.instruction()->name() << " instruction" << std::endl
+           << "Substituted operands: " << subs_ops_mnems.str() << std::endl
+           << "Instruction format: "
+           << context::instruction_sets::mach_format_to_string(mnemonic_instr.instruction()->format());
+    items.emplace(std::string(mnemonic_instr.name()),
+        detail_ss.str(),
+        std::string(mnemonic_instr.name()) + "   " + subs_ops_nomnems.str(),
+        doc_ss.str(),
+        completion_item_kind::mach_instr);
+}
 
+void process_ca_instruction(const context::ca_instruction& ca_instr, instruction_completion_items::storage& items)
+{
+    items.emplace(std::string(ca_instr.name()),
+        "",
+        std::string(ca_instr.name()),
+        "Conditional Assembly",
+        completion_item_kind::ca_instr);
+}
+
+struct instruction_visitor
+{
+    instruction_visitor(instruction_completion_items::storage& items)
+        : items(items) {};
+
+    void operator()(const context::ca_instruction* ca_instr)
+    {
+        if (ca_instr)
+            process_ca_instruction(*ca_instr, items);
+    };
+    void operator()(const context::assembler_instruction* asm_instr)
+    {
+        if (asm_instr)
+            process_assembler_instruction(*asm_instr, items);
+    };
+    void operator()(const context::machine_instruction* mach_instr)
+    {
+        if (mach_instr)
+            process_machine_instruction(*mach_instr, items);
+    };
+    void operator()(const context::mnemonic_code* mne_code)
+    {
+        if (mne_code)
+            process_mnemonic_code(*mne_code, items);
+    };
+    void operator()(auto) { /* Do nothing */ };
+
+    std::set<completion_item_s, completion_item_s::label_comparer>& items;
+};
+
+
+const instruction_completion_items::storage& get_completion_items(
+    const std::unordered_map<context::id_index, context::opcode_t::opcode_variant>& instr_map)
+{
+    using namespace context;
 
+    static instruction_completion_items::storage items;
+
+    if (!items.empty())
+    {
+        return items;
+    }
+
+    auto visitor = instruction_visitor { items };
+
+    for (const auto& [key, val] : instr_map)
+    {
+        std::visit(visitor, val);
+    }
+
+    return items;
+}
+} // namespace
+
+instruction_completion_items::instruction_completion_items(
```

Comment:
wouldn't be easier to just have one static full list here and then filter it in `lsp_context::complete_instr`?